core/matrix/fbcsr.cpp

```diff
@@ -0,0 +1,504 @@
+            } else {
+                // If this does not happen, we re-visited a nonzero
+                assert(nnzblk(localrow, localcol) == gko::zero<value_type>());
+                nnzblk(localrow, localcol) = val;
+            }
+        }
+        return blocks;
+    };
+
+    const std::map<FbEntry, Block_t, FbLess> blocks = create_block_map(data);
+
+    auto tmp = Fbcsr::create(this->get_executor()->get_master(), data.size,
+                             blocks.size() * bs * bs, bs);
+
+    tmp->row_ptrs_.get_data()[0] = 0;
+    index_type cur_brow = 0;
+    index_type cur_bnz = 0;
+    index_type cur_bcol = blocks.begin()->first.block_column;
+    const index_type num_brows = data.size[0] / bs;
+
+    blockutils::DenseBlocksView<value_type, index_type> values(
+        tmp->values_.get_data(), bs, bs);
+
+    for (auto it = blocks.begin(); it != blocks.end(); it++) {
+        GKO_ENSURE_IN_BOUNDS(cur_brow, num_brows);
+
+        tmp->col_idxs_.get_data()[cur_bnz] = it->first.block_column;
+        for (int ibr = 0; ibr < bs; ibr++) {
+            for (int jbr = 0; jbr < bs; jbr++) {
+                values(cur_bnz, ibr, jbr) = it->second(ibr, jbr);
+            }
+        }
+        if (it->first.block_row > cur_brow) {
+            tmp->row_ptrs_.get_data()[++cur_brow] = cur_bnz;
+        } else {
+            assert(cur_brow == it->first.block_row);
+            assert(cur_bcol <= it->first.block_column);
+        }
+
+        cur_bcol = it->first.block_column;
+        cur_bnz++;
+    }
+
+    tmp->row_ptrs_.get_data()[++cur_brow] =
+        static_cast<index_type>(blocks.size());
+
+    assert(cur_brow == tmp->get_size()[0] / bs);
+
+    tmp->move_to(this);
+}
+
+
+template <typename ValueType, typename IndexType>
+void Fbcsr<ValueType, IndexType>::write(mat_data &data) const
+{
+    std::unique_ptr<const LinOp> op{};
+    const Fbcsr *tmp{};
+    if (this->get_executor()->get_master() != this->get_executor()) {
+        op = this->clone(this->get_executor()->get_master());
+        tmp = static_cast<const Fbcsr *>(op.get());
+    } else {
+        tmp = this;
+    }
+
+    data = {tmp->get_size(), {}};
+
+    const blockutils::DenseBlocksView<const value_type, index_type> vblocks(
+        tmp->values_.get_const_data(), bs_, bs_);
+
+    for (size_type brow = 0; brow < tmp->get_num_block_rows(); ++brow) {
+        const auto start = tmp->row_ptrs_.get_const_data()[brow];
+        const auto end = tmp->row_ptrs_.get_const_data()[brow + 1];
+
+        for (auto inz = start; inz < end; ++inz) {
+            for (int ib = 0; ib < bs_; ib++) {
+                const auto row = brow * bs_ + ib;
+                for (int jb = 0; jb < bs_; jb++) {
+                    const auto col =
+                        tmp->col_idxs_.get_const_data()[inz] * bs_ + jb;
+                    const auto val = vblocks(inz, ib, jb);
+                    data.nonzeros.emplace_back(row, col, val);
+                }
```

Comment:
Although our documentation doesn't guarantee any particular order on matrix_data.nonzeros, all of our other matrix formats produce ascending row indices. This loop order causes the output to instead only have ascending block row indices. Swapping the nonzero and block-local row loop should fix that.